[BUG] Databricks nightly CI: test_buckets OOM failure (CPU) on DB 17.3

**Describe the bug**
Build: rapids_databricks_nightly-dev-github/2175 (DB_RUNTIME=17.3)

The Databricks nightly CI test run for DB_RUNTIME=17.3 (Spark 4.0.0) failed with 1 test failure out of 44555 tests. The failing test `test_buckets[parquet-reader_confs9]` encountered a `SparkOutOfMemoryError` **during the CPU run phase**. A native crash with `std::bad_alloc` and heap corruption signal was also observed, indicating memory exhaustion after ~2.5 hours of parallel testing. A secondary failure occurred because the eventlog tar command used an incorrect path (`integration_tests/target/run_dir*/` instead of `scala2.13/integration_tests/target/run_dir*/`).

Error logs:
```
FAILED parquet_test.py::test_buckets[parquet-reader_confs9] - py4j.protocol.Py4JJavaError:
org.apache.spark.SparkException: Job aborted due to stage failure: Task 3 in stage 19775.0 failed 1 times
Caused by: org.apache.spark.memory.SparkOutOfMemoryError: [UNABLE_TO_ACQUIRE_MEMORY] Unable to acquire 65536 bytes of memory, got 0. SQLSTATE: 53200
	at org.apache.spark.memory.MemoryConsumer.throwOom(MemoryConsumer.java:187)
	at org.apache.spark.util.collection.unsafe.sort.UnsafeExternalSorter.growPointerArrayIfNecessary

libc++abi: terminating due to uncaught exception of type std::bad_alloc: std::bad_alloc
Received signal 6: Aborted
@0x00007f9804a0c330 malloc_consolidate(): unaligned fastbin chunk detected

tar: integration_tests/target/run_dir*: Cannot stat: No such file or directory
```

**Environment details**
- Databricks Runtime: 17.3.x-gpu-ml-scala2.13 (Spark 4.0.0)
- Test mode: CI_PART1
- Test duration: ~2.5 hours with 8 parallel workers
- 1 failed, 39624 passed, 2812 skipped out of 44555 items

```
[2026-04-09T18:17:01.922Z] ++ cat /databricks/BUILDINFO
[2026-04-09T18:17:01.922Z] BUILD_SCM_BRANCH HEAD
[2026-04-09T18:17:01.922Z] BUILD_SCM_REVISION ed3c5fc1e748977a7a26c0eb2d56894e3d7694c4
[2026-04-09T18:17:01.922Z] BUILD_SCM_SHORT_HASH ed3c5fc
[2026-04-09T18:17:01.922Z] BUILD_SCM_STATUS Clean
[2026-04-09T18:17:01.922Z] BUILD_TIMESTAMP 1775514701
[2026-04-09T18:17:01.922Z] DATEHASH 20260406153141-ed3c5fc1e748977a7a26c0eb2d56894e3d7694c4
[2026-04-09T18:17:01.922Z] FORMATTED_DATE 2026 Apr 06 22 31 41 Mon
```

--- a/integration_tests/src/main/python/parquet_test.py
+++ b/integration_tests/src/main/python/parquet_test.py
@@ -884,12 +884,12 @@ _BUCKET_TEST_LEFT_ROWS = 100_000
 _BUCKET_TEST_RIGHT_ROWS = 1_000_000
 
 def createBucketedTableAndJoin(spark, tbl_1, tbl_2):
-    # Keep this large enough to exercise bucketed joins, but small enough that the CPU
-    # sort/shuffle path stays stable when all reader configs run concurrently in CI.
-    spark.range(_BUCKET_TEST_LEFT_ROWS).write.bucketBy(_BUCKET_TEST_NUM_BUCKETS, "id") \
-        .sortBy("id").mode('overwrite').saveAsTable(tbl_1)
-    spark.range(_BUCKET_TEST_RIGHT_ROWS).write.bucketBy(_BUCKET_TEST_NUM_BUCKETS, "id") \
-        .sortBy("id").mode('overwrite').saveAsTable(tbl_2)
+    # Keep this large enough to exercise bucketed joins, while reducing CPU-side sort
+    # pressure. Both sides use the same bucket count so the bucketed join still applies.
+    (spark.range(_BUCKET_TEST_LEFT_ROWS).write.bucketBy(_BUCKET_TEST_NUM_BUCKETS, "id")
+        .sortBy("id").mode('overwrite').saveAsTable(tbl_1))
+    (spark.range(_BUCKET_TEST_RIGHT_ROWS).write.bucketBy(_BUCKET_TEST_NUM_BUCKETS, "id")
+        .sortBy("id").mode('overwrite').saveAsTable(tbl_2))
     bucketed_left = spark.table(tbl_1)
     bucketed_right = spark.table(tbl_2)
     return bucketed_left.join(bucketed_right, "id")

--- a/integration_tests/src/main/python/parquet_test.py
+++ b/integration_tests/src/main/python/parquet_test.py
@@ -879,12 +879,20 @@ def test_parquet_input_meta_fallback(spark_tmp_path, v1_enabled_list, reader_con
                         'input_file_block_length()'),
             conf=all_confs)
 
+_BUCKET_TEST_NUM_BUCKETS = 8
+_BUCKET_TEST_LEFT_ROWS = 100_000
+_BUCKET_TEST_RIGHT_ROWS = 1_000_000
+
 def createBucketedTableAndJoin(spark, tbl_1, tbl_2):
-    spark.range(10e4).write.bucketBy(4, "id").sortBy("id").mode('overwrite').saveAsTable(tbl_1)
-    spark.range(10e6).write.bucketBy(4, "id").sortBy("id").mode('overwrite').saveAsTable(tbl_2)
-    bucketed_4_10e4 = spark.table(tbl_1)
-    bucketed_4_10e6 = spark.table(tbl_2)
-    return bucketed_4_10e4.join(bucketed_4_10e6, "id")
+    # Keep this large enough to exercise bucketed joins, but small enough that the CPU
+    # sort/shuffle path stays stable when all reader configs run concurrently in CI.
+    spark.range(_BUCKET_TEST_LEFT_ROWS).write.bucketBy(_BUCKET_TEST_NUM_BUCKETS, "id") \
+        .sortBy("id").mode('overwrite').saveAsTable(tbl_1)
+    spark.range(_BUCKET_TEST_RIGHT_ROWS).write.bucketBy(_BUCKET_TEST_NUM_BUCKETS, "id") \
+        .sortBy("id").mode('overwrite').saveAsTable(tbl_2)
+    bucketed_left = spark.table(tbl_1)
+    bucketed_right = spark.table(tbl_2)
+    return bucketed_left.join(bucketed_right, "id")
 
 @ignore_order
 @allow_non_gpu('DataWritingCommandExec,ExecutedCommandExec,WriteFilesExec')

--- a/integration_tests/src/main/python/parquet_test.py
+++ b/integration_tests/src/main/python/parquet_test.py
@@ -879,13 +879,14 @@ def test_parquet_input_meta_fallback(spark_tmp_path, v1_enabled_list, reader_con
                         'input_file_block_length()'),
             conf=all_confs)
 
+# More buckets reduce per-task CPU sort pressure. Both sides must use the same
+# bucket count so Spark can still use the bucketed join path.
 _BUCKET_TEST_NUM_BUCKETS = 8
 _BUCKET_TEST_LEFT_ROWS = 100_000
 _BUCKET_TEST_RIGHT_ROWS = 1_000_000
 
 def createBucketedTableAndJoin(spark, tbl_1, tbl_2):
-    # Keep this large enough to exercise bucketed joins, while reducing CPU-side sort
-    # pressure. Both sides use the same bucket count so the bucketed join still applies.
+    # Keep this large enough to exercise bucketed joins, while reducing CPU-side sort pressure.
     (spark.range(_BUCKET_TEST_LEFT_ROWS).write.bucketBy(_BUCKET_TEST_NUM_BUCKETS, "id")
         .sortBy("id").mode('overwrite').saveAsTable(tbl_1))
     (spark.range(_BUCKET_TEST_RIGHT_ROWS).write.bucketBy(_BUCKET_TEST_NUM_BUCKETS, "id")

--- a/jenkins/databricks/run-tests.py
+++ b/jenkins/databricks/run-tests.py
@@ -22,9 +22,15 @@ import params
 
 
 def get_test_report_path_prefix():
-    """Return the remote repo root whose integration_tests/target contains run_dir outputs."""
+    """Return the remote repo root whose integration_tests/target contains run_dir outputs.
+
+    Spark 4+ Databricks builds run from scala2.13/pom.xml, so run_pyspark_from_build.sh
+    writes run_dir* under spark-rapids/scala2.13/integration_tests/target. Spark 3.x
+    builds still use the root pom and write under spark-rapids/integration_tests/target.
+    """
     source_path = (params.jar_path if params.jar_path else "/home/ubuntu/spark-rapids").rstrip("/")
-    if params.base_spark_pom_version.startswith("4.") and not source_path.endswith("/scala2.13"):
+    if int(params.base_spark_pom_version.split(".", 1)[0]) >= 4 and \
+            not source_path.endswith("/scala2.13"):
         return "%s/scala2.13" % source_path
     return source_path
 

--- a/jenkins/databricks/run-tests.py
+++ b/jenkins/databricks/run-tests.py
@@ -21,6 +21,14 @@ from clusterutils import ClusterUtils
 import params
 
 
+def get_test_report_path_prefix():
+    """Return the remote repo root whose integration_tests/target contains run_dir outputs."""
+    source_path = (params.jar_path if params.jar_path else "/home/ubuntu/spark-rapids").rstrip("/")
+    if params.base_spark_pom_version.startswith("4.") and not source_path.endswith("/scala2.13"):
+        return "%s/scala2.13" % source_path
+    return source_path
+
+
 def main():
     """Define main function."""
     master_addr = ClusterUtils.cluster_get_master_addr(params.workspace, params.clusterid, params.token)
@@ -45,16 +53,17 @@ def main():
     try:
         subprocess.check_call(ssh_command, shell=True)
     finally:
-        print("Copying test report tarball back")
-        if params.base_spark_pom_version.startswith("4."):
-            default_jar_path = "/home/ubuntu/spark-rapids/scala2.13"
-        else:
-            default_jar_path = "/home/ubuntu/spark-rapids"
-        report_path_prefix = params.jar_path if params.jar_path else default_jar_path
-        rsync_command = "rsync -I -Pave \"ssh %s\" ubuntu@%s:%s/integration_tests/target/run_dir*/TEST-pytest-*.xml ./" % \
-            (ssh_args, master_addr, report_path_prefix)
+        print("Copying test reports back")
+        report_target_path = "%s/integration_tests/target" % get_test_report_path_prefix()
+        subprocess.check_call("mkdir -p integration_tests/target", shell=True)
+        rsync_command = "rsync -I -Pave \"ssh %s\" ubuntu@%s:%s/run_dir* integration_tests/target/" % \
+            (ssh_args, master_addr, report_target_path)
         print("rsync command: %s" % rsync_command)
-        subprocess.check_call(rsync_command, shell = True)
+        subprocess.check_call(rsync_command, shell=True)
+        rsync_command = "rsync -I -Pave \"ssh %s\" ubuntu@%s:%s/integration_tests/target/run_dir*/TEST-pytest-*.xml ./" % \
+            (ssh_args, master_addr, get_test_report_path_prefix())
+        print("rsync command: %s" % rsync_command)
+        subprocess.check_call(rsync_command, shell=True)
 
 
 if __name__ == '__main__':

--- a/jenkins/databricks/run-tests.py
+++ b/jenkins/databricks/run-tests.py
@@ -24,13 +24,16 @@ import params
 def get_test_report_path_prefix():
     """Return the remote repo root whose integration_tests/target contains run_dir outputs.
 
-    Spark 4+ Databricks builds run from scala2.13/pom.xml, so run_pyspark_from_build.sh
-    writes run_dir* under spark-rapids/scala2.13/integration_tests/target. Spark 3.x
-    builds still use the root pom and write under spark-rapids/integration_tests/target.
+    When jar_path is provided, tests run from that uploaded JAR directory. Otherwise Spark
+    4+ Databricks builds run from scala2.13/pom.xml, so run_pyspark_from_build.sh writes
+    run_dir* under spark-rapids/scala2.13/integration_tests/target.
     """
-    source_path = (params.jar_path if params.jar_path else "/home/ubuntu/spark-rapids").rstrip("/")
-    if int(params.base_spark_pom_version.split(".", 1)[0]) >= 4 and \
-            not source_path.endswith("/scala2.13"):
+    source_path = (params.jar_path or "/home/ubuntu/spark-rapids").rstrip("/")
+    if params.jar_path:
+        return source_path
+
+    spark_major_version = int(params.base_spark_pom_version.split(".", 1)[0])
+    if spark_major_version >= 4:
         return "%s/scala2.13" % source_path
     return source_path
 

--- a/jenkins/databricks/run-tests.py
+++ b/jenkins/databricks/run-tests.py
@@ -62,14 +62,16 @@ def main():
         print("Copying test reports back")
         report_target_path = "%s/integration_tests/target" % get_test_report_path_prefix()
         subprocess.check_call("mkdir -p integration_tests/target", shell=True)
+        # Copy the full run directories because they contain more diagnostics than the
+        # JUnit XML files, such as Spark event logs and worker logs.
         rsync_command = "rsync -I -Pave \"ssh %s\" ubuntu@%s:%s/run_dir* integration_tests/target/" % \
             (ssh_args, master_addr, report_target_path)
         print("rsync command: %s" % rsync_command)
         subprocess.check_call(rsync_command, shell=True)
-        rsync_command = "rsync -I -Pave \"ssh %s\" ubuntu@%s:%s/integration_tests/target/run_dir*/TEST-pytest-*.xml ./" % \
-            (ssh_args, master_addr, get_test_report_path_prefix())
-        print("rsync command: %s" % rsync_command)
-        subprocess.check_call(rsync_command, shell=True)
+        # Keep the existing Jenkins JUnit publishing flow, which expects XML files in this directory.
+        copy_xml_command = "cp integration_tests/target/run_dir*/TEST-pytest-*.xml ./"
+        print("copy xml command: %s" % copy_xml_command)
+        subprocess.check_call(copy_xml_command, shell=True)
 
 
 if __name__ == '__main__':

--- a/jenkins/databricks/run-tests.py
+++ b/jenkins/databricks/run-tests.py
@@ -69,7 +69,7 @@ def main():
         print("rsync command: %s" % rsync_command)
         subprocess.check_call(rsync_command, shell=True)
         # Keep the existing Jenkins JUnit publishing flow, which expects XML files in this directory.
-        copy_xml_command = "cp integration_tests/target/run_dir*/TEST-pytest-*.xml ./"
+        copy_xml_command = "cp integration_tests/target/run_dir*/TEST-pytest-*.xml ./ || true"
         print("copy xml command: %s" % copy_xml_command)
         subprocess.check_call(copy_xml_command, shell=True)
 

--- a/jenkins/databricks/run-tests.py
+++ b/jenkins/databricks/run-tests.py
@@ -32,7 +32,12 @@ def get_test_report_path_prefix():
     if params.jar_path:
         return source_path
 
-    spark_major_version = int(params.base_spark_pom_version.split(".", 1)[0])
+    try:
+        spark_major_version = int(params.base_spark_pom_version.split(".", 1)[0])
+    except ValueError:
+        print("WARNING: unable to parse base Spark version '%s'; using the default report path" %
+              params.base_spark_pom_version)
+        spark_major_version = 0
     if spark_major_version >= 4:
         return "%s/scala2.13" % source_path
     return source_path
@@ -63,18 +68,26 @@ def main():
         subprocess.check_call(ssh_command, shell=True)
     finally:
         print("Copying test reports back")
-        report_target_path = "%s/integration_tests/target" % get_test_report_path_prefix()
-        subprocess.check_call("mkdir -p integration_tests/target", shell=True)
-        # Copy the full run directories because they contain more diagnostics than the
-        # JUnit XML files, such as Spark event logs and worker logs.
-        rsync_command = "rsync -I -Pave \"ssh %s\" ubuntu@%s:%s/run_dir* integration_tests/target/" % \
-            (ssh_args, master_addr, report_target_path)
-        print("rsync command: %s" % rsync_command)
-        subprocess.check_call(rsync_command, shell=True)
-        # Keep the existing Jenkins JUnit publishing flow, which expects XML files in this directory.
-        copy_xml_command = "cp integration_tests/target/run_dir*/TEST-pytest-*.xml ./ || true"
-        print("copy xml command: %s" % copy_xml_command)
-        subprocess.check_call(copy_xml_command, shell=True)
+        try:
+            report_target_path = "%s/integration_tests/target" % get_test_report_path_prefix()
+            subprocess.check_call("mkdir -p integration_tests/target", shell=True)
+            # Copy the diagnostics needed by Jenkins while avoiding unrelated run_dir contents.
+            rsync_command = "rsync -I -Pave \"ssh %s\" --prune-empty-dirs " \
+                "--include='run_dir*/' " \
+                "--include='run_dir*/TEST-pytest-*.xml' " \
+                "--include='run_dir*/eventlog_*/***' " \
+                "--include='run_dir*/*_worker_logs.log' " \
+                "--exclude='*' ubuntu@%s:%s/ integration_tests/target/" % \
+                (ssh_args, master_addr, report_target_path)
+            print("rsync command: %s" % rsync_command)
+            subprocess.check_call(rsync_command, shell=True)
+            # Keep the existing Jenkins JUnit publishing flow, which expects XML files in this
+            # directory.
+            copy_xml_command = "cp integration_tests/target/run_dir*/TEST-pytest-*.xml ./ || true"
+            print("copy xml command: %s" % copy_xml_command)
+            subprocess.check_call(copy_xml_command, shell=True)
+        except subprocess.CalledProcessError as e:
+            print("WARNING: test report collection failed: %s" % e)
 
 
 if __name__ == '__main__':

